File: src/easy/Q0680.py
"""
680. Valid Palindrome II
https://leetcode.com/problems/valid-palindrome-ii/
"""


# Time: O(n), Space: O(1), where n is the length of `s`
def valid_palindrome(s: str) -> bool:
    result = True
    lo, hi = 0, len(s) - 1

    # Check if `s` is a palindrome without removing any characters
    while lo < hi:
        if s[lo] != s[hi]:
            break
        lo += 1
        hi -= 1

    old_lo = lo
    old_hi = hi

    # Check if removal of `s[lo]` makes `s` a palindrome
    result_lo = True
    lo += 1
    while lo < hi:
        if s[lo] != s[hi]:
            result_lo = False
            break
        lo += 1
        hi -= 1

    # Reset `lo` and `hi`
    lo = old_lo
    hi = old_hi

    # Check if removal of `s[hi]` makes `s` a palindrome
    result_hi = True
    hi -= 1
    while lo < hi:
        if s[lo] != s[hi]:
            result_hi = False
            break
        lo += 1
        hi -= 1

    return result and (result_lo or result_hi)

File: src/easy/Q0680/Q0680.cpp
// 680. Valid Palindrome II
// https://leetcode.com/problems/valid-palindrome-ii/

#include <string>

using namespace std;

// Time: O(n), Space: O(1), where n is the size of `s`.
bool valid_palindrome(string s) {
  bool result1 = true;
  size_t lo = 0;
  size_t hi = s.size() - 1;
  while (lo < hi) {
    if (s[lo] != s[hi]) {
      result1 = false;
      break;
    }
    ++lo;
    --hi;
  }

  size_t cached_lo = lo;
  size_t cached_hi = hi;

  bool result2 = true;
  ++lo;
  while (lo >= 0 and hi < s.size() and lo < hi) {
    if (s[lo] != s[hi]) {
      result2 = false;
      break;
    }
    ++lo;
    --hi;
  }

  lo = cached_lo;
  hi = cached_hi;

  bool result3 = true;
  --hi;
  while (lo >= 0 and hi < s.size() and lo < hi) {
    if (s[lo] != s[hi]) {
      result3 = false;
      break;
    }
    ++lo;
    --hi;
  }

  return result1 or result2 or result3;
}
